### PR TITLE
WIP: Solana integration

### PR DESCRIPTION
**Pull request by brewmaster012**
# Description

This PR adds level 1 support for Solana blockchain; namely the following features: 
1. deposit SOL and SPL tokens into ZetaChain as ZRC20
2. deposit and call zEVM contract from Solana
3. withdraw SOL and SPL tokens from ZRC20

On the Solana network a program (smart contract) manages authentication, custody, and remote call:
https://github.com/zeta-chain/protocol-contracts-solana

Closes: <PD-XXXX>

# Tasks
- [x] Initialize the Gateway contract on localnet e2e test Solana blockchain
- [x] Deposit SOL into ZRC20
- [ ] Withdraw SOL from ZRC20
- [ ] Deposit SPL Token into ZRC20 
- [ ] Withdraw SPL Token from ZRC20
- [ ] Desposit SOL and call (swap Ether?)
- [ ] Deposit SPL Token and call (swap Ether?)
- [ ] Unit tests 
- [ ] Remove all hardcoded localnet settings in generic code paths

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [x] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: contrib/localnet/solana/start-solana.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+solana config set --url localhost
+echo "starting solana test validator..."
+solana-test-validator &
+
+sleep 5
+# airdrop to e2e sol account
+solana airdrop 100
+solana airdrop 100 37yGiHAnLvWZUNVwu9esp74YQFqxU1qHCbABkDvRddUQ
+solana program deploy gateway.so
```
Comment (gartnera) [inline contrib/localnet/solana/start-solana.sh:14]: The deployment should be done in `zetae2e` and the funding should be done in `zetae2e-setup.sh`. I imagine the airdrop rpc works remotely right? You'd need to copy the `solana` binary into the e2e orchestrator container

File: contrib/localnet/solana/Dockerfile.solana
[the remark this answers is not in the document]
Comment (gartnera) [reply contrib/localnet/solana/Dockerfile.solana:1]: It looks like the `solana-test-validator` is pretty cpu intensive, so we might not want to run it all the time

**Review by lumtis — COMMENTED** [review]
@brewmaster012 what is `cmd/solana/main.go`?
Is it the logic that simulate what ZetaClient should do?

**Comment by brewmaster012** [comment]
> @brewmaster012 what is `cmd/solana/main.go`? Is it the logic that simulate what ZetaClient should do?

yes, just for experimentation.

**Comment by lumtis** [comment]
Closing as withdraw and deposit have been implemented

**Pull request closed without merging**